lib/basedriver/driver.js

```diff
@@ -408,6 +423,31 @@ class BaseDriver extends Protocol {
     return false;
   }
 
+  proxyRouteIsAvoided (sessionId, method, url) {
+    for (let avoidSchema of this.getProxyAvoidList(sessionId)) {
+      if (!_.isArray(avoidSchema) || avoidSchema.length !== 2) {
+        throw new Error('Proxy avoidance must be a list of pairs');
+      }
+      let [avoidMethod, avoidPathRegex] = avoidSchema;
+      if (!_.includes(['GET', 'POST', 'DELETE'], avoidMethod)) {
+        throw new Error(`Unrecognized proxy avoidance method '${avoidMethod}'`);
+      }
+      if (!(avoidPathRegex instanceof RegExp)) {
+        throw new Error('Proxy avoidance path must be a regular expression');
+      }
+      let normalizedUrl = url.replace(/^\/wd\/hub/, '');
+      if (avoidMethod === method && avoidPathRegex.test(normalizedUrl)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  async proxyCommand (/*url, method, body*/) {
```

Comment:
why do we need to implement it in each driver?

Reply:
each driver does proxying differently (uiauto2 calls `this.uiautomator2.jwproxy` for example, which isn't the case with xcui)

Reply:
cannot we just call `click` or `performActions`? Why the base driver should care about proxying?

Reply:
what if a driver doesn't implement `performActions` itself, but merely proxies requests for it to a downstream driver? for example, if uiauto2 server implemented the /actions endpoint, but uiauto2 driver did not. this is a totally valid case. we have many endpoints which are not specified explicitly in a driver and instead we rely on proxying to access the endpoint on a downstream driver.

does that make sense? maybe i'm confused but i think it's a real issue (at least in principle)

Reply:
I wouldn't try to outsmart drivers. If some endpoint is not implemented there and it exists in the downstream driver then it would be better and more practical to implement the missing stuff directly there. Adding new requirements to base driver creates unnecessary complexity for the whole infrastructure.

Reply:
ok. i think you have a good point. the way i am doing it now we will also require drivers to implement the `proxyCommand` method. so it's not less onerous than requiring them to simply implement `performActions` and `performTouch`. i'll do the same thing as i have done elsewhere and simply throw a meaningful error if these methods do not exist on the driver.

if the driver typically proxies the action/touch commands, then it will need to basically implement a one-line function to support image elements, like:

```js
commands.performActions = function (actions) {
  this.proxyCommand('POST', '/actions', actions);
};
```

In most cases this will be bypassed, but in the case of ImageElement actions, it will be called.